demos/test_vlc_net/src/server.rs

```diff
@@ -0,0 +1,845 @@
+            state.metrics.clock_update_count = state.metrics.clock_update_count + 1;
+            let mut event = P2PMsg {
+                clock: state.clock.clone(),
+                data: payload_data.clone(),
+                pub_key: None,
+                signature: None,
+            };
+
+            // Serialize and sign the event
+            let serialized_event = match bincode::options().serialize(&event) {
+                Ok(vec) => vec,
+                Err(e) => {
+                    error!("Failed to serialize event: {}", e);
+                    return None;
+                }
+            };
+
+            let signature = match self.local_key.sign(&serialized_event) {
+                Ok(sig) => sig,
+                Err(e) => {
+                    error!("Failed to sign event: {}", e);
+                    return None;
+                }
+            };
+
+            event.signature = Some(signature);
+            event.pub_key = Some(pub_key.clone());
+            let message = match bincode::options().serialize(&event) {
+                Ok(msg) => msg,
+                Err(e) => {
+                    error!("Failed to serialize signed event: {}", e);
+                    return None;
+                }
+            };
+            let message_len = message.len();
+            (event, message, message_len, send_counts)
+        };
+        Some((event, message, message_len, send_counts))
+    }
+
+    async fn event_handler(
+        self: Arc<Self>,
+        mut rx_net: mpsc::Receiver<Event>,
+        tx_async_state: mpsc::Sender<StateUpdate>,
+    ) {
+        if self.config.bootnodes.is_none() {
+            info!("bootnode don't parse events");
+            return;
+        }
+        let semaphore = Arc::new(Semaphore::new(self.config.concurrent_verify as usize));
+
+        // Event loop to handle incoming messages
+        loop {
+            let event = match rx_net.recv().await {
+                Some(ev) => ev,
+                None => break, // Receiver has been closed
+            };
+
+            match event {
+                Event::PeerConnected(peer_id) => {
+                    debug!("connected to {:?}", peer_id);
+                }
+                Event::PeerSubscribed(peer_id, topic_hash) => {
+                    debug!("Remote peer {:?} subscribed to {:?}", peer_id, topic_hash);
+                }
+                Event::Message(msg) => {
+                    // self.clone().handle_p2p_msg(msg, tx_async_state.clone()).await;
+                    let self_clone = self.clone();
+                    let tx_state_clone = tx_async_state.clone();
+                    if let Ok(_permit) = semaphore.try_acquire() {
+                        tokio::spawn(async move {
+                            self_clone.handle_p2p_msg(msg, tx_state_clone).await;
+                        });
+                    } else {
+                        warn!("Too many concurrent tasks, dropping message.");
```

Comment:
When the semaphore is full, messages are simply dropped
This will lead to loss of messages and inconsistent state

Reply:
This is a trick for addressing high-message press situations, the only way is to drop the messages when low-performance machine and high network overhead.

Reply:
Should not drop messages in any condition, unless it's illegal

Reply:
This can be left to be discussed at the meeting.